urlcaef331f4e_dreamable_cra_tgz-pJ8-coho_lp_project_LPProjectJ8/src/coho/geom/twodim/SimplePolygon.java

```diff
@@ -0,0 +1,676 @@
+		}else{
+			//if(points==null || points.length==0){
+			if(points==null || points.length <1 ){
+				//throw new GeomException("null polygon");
+				this.points = new Point[0];
+				this.type = CohoDouble.type; //NOTE Let matlab handle it
+			}else{
+				//Use hashset rather than contain() of ArrayList to save time
+				HashSet<Point> set = new HashSet<Point>();
+				ArrayList<Point> list = new ArrayList<Point>();
+				this.type = points[0].type();
+				for(int i=0; i<points.length; i++){
+					if(set.add(points[i])){//copy and remove duplicated points
+						if(type!=points[i].type())
+							throw new GeomException("Type for polygon is not the same");
+						list.add(points[i]);
+					}else{
+						//System.out.println("repeated point "+points[i]+"found");
+					}
+				}
+				if(list.size()<3){
+					throw new GeomException("A polygon must have at least three vertices");
+				}
+				this.points = list.toArray(new Point[list.size()]);
+			}
+		}
+	}
+	
+	//NOTICE: the direction of segment must be anti-clock-wise.
+	public SimplePolygon(Segment[] segs){
+		Point curr;
+		try{
+			this.type = segs[0].type();
+			Hashtable<Point,Point> map  = new Hashtable<Point,Point>();
+			for(int i=0; i<segs.length; i++){
+				if(type!=segs[i].type())
+					throw new GeomException("type for polygon is not the same");
+				map.put(segs[i].p(0), segs[i].p(1));
+			}
+			points = new Point[segs.length];
+			curr = segs[0].p(0);
+			for(int i=0; i<segs.length;i++){
+				points[i] = curr;
+				curr = map.get(curr);
+			}
+		}catch(Exception e){
+			throw new GeomException("The polygon is invalid");
+		}
+		if(!curr.equals(segs[0].p(0))){
+			throw new GeomException("The polygon is not closed up");
+		}
+	}
+	
+	public int degree(){
+		return points.length;
+	}
+	public Point[] points(){
+		return points.clone();
+	}
+	public Point point(int i){//i should be 0-degree()-1
+		return points[i];
+	}
+	public Segment[] edges(){
+		Segment[] edges = new Segment[points.length];
+		for(int i=0; i<points.length; i++){
+			edges[i] = Segment.create(points[i], points[(i+1)%points.length]);
+		}			
+		return edges;
+	}
+	public Segment edge(int i){
+		return Segment.create(points[i], points[(i+1)%points.length]);
+	}
+	
+	/**
+	 * Similar with points, but it start from the ll point;
+	 * TODO: return ordered points and edges
+	 */
+	public Point[] vertices(){
+		int llIndex = llPos(), n = points.length;
+		ArrayList<Point> list = new ArrayList<Point>();
+		for(int i=0; i<n; i++){
+			list.add(points[(i+llIndex)%n]);
+		}
+		return list.toArray(new Point[list.size()]);
+	}
+	
+	public int llPos(){
+		int llIndex = 0, n = points.length;
+		Point ll = points[0];
+		for(int i=1; i<n; i++){
+			if(points[i].compareTo(ll)<0){
+				ll = points[i];
+				llIndex = i;
+			}
+		}
+		return llIndex;
+	}
+	
+	public Point ll(){
+		Point ll = points[0];
+		for(int i=1; i<points.length; i++){
+			if(points[i].compareTo(ll)<0)
+				ll = points[i];
+		}
+		return ll;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 156:161, 164:169.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.